### PR TITLE
UINV-101 Create settings form for Batch group settings (Multiple Batch groups)

### PR DESCRIPTION
**Pull request by AndreiShumski2**
## Purpose
Create settings form for Batch group settings (Multiple Batch groups)
https://issues.folio.org/browse/UINV-101
## Screenshot
image
image

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.
- [x] I've added appropriate record to the CHANGELOG.md
- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] If any API-related changes - okapi interfaces and permissions are reviewed/changed correspondingly
  - [x] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js [hunk 1]
```diff
@@ -1,7 +1,86 @@
...
 
-const BatchGroupConfigurationSettings = () => {
-  return <div data-test-batch-group-configuration-settings />;
+import {
+  Pane,
+  Spinner,
+} from '@folio/stripes/components';
+
+import { batchGroupsResource } from '../../common/resources';
+import BatchGroupConfigurationSettingsForm from './BatchGroupCongiurationSettingsForm';
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: IMHO, word `Settings` could be omitted here without consequences

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js [hunk 2]
```diff
@@ -1,7 +1,86 @@
...
+
+const BatchGroupConfigurationSettings = ({ mutator }) => {
+  const [batchGroups, setBatchGroups] = useState();
+  const [batchGroupId, setBatchGroupId] = useState();
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: `batchGroupId` - it's good, I know that you put id of batch group there, but for what purpose? Maybe it's a good idea to add a purpose word, something like selectedBatchGroupId or else

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js [hunk 3]
```diff
@@ -1,7 +1,86 @@
...
+
+  const batchGroupsOptions = useMemo(() => {
+    if (!batchGroups) return [];
+
+    return batchGroups.map(({ name, id }) => ({ label: name, value: id })) || [];
+  }, [batchGroups]);
+
+  const onSaveExportConfig = useCallback(
+    () => {
+
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: is it supposed to have empty body?

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js [hunk 4]
```diff
@@ -1,7 +1,86 @@
...
+  // const onChangeBatchGroup = useCallback((id) => setBatchGroupId(id), []);
+
+  if (!(batchGroups && batchGroupId)) {
+    return (
+      <Pane
+        defaultWidth="fill"
+        paneTitle={paneTitle}
+      >
+        <Spinner />
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: do we have any guideline to add spinners? I', not aware of any

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js [hunk 5]
```diff
@@ -1,7 +1,86 @@
...
+  const onSaveExportConfig = useCallback(
+    () => {
+
+    },
+    [batchGroupId],
+  );
+
+  // const onChangeBatchGroup = useCallback((id) => setBatchGroupId(id), []);
+
+  if (!(batchGroups && batchGroupId)) {
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: what if the list is empty, you'll always show spinner? IMHO, it's not the purpose of spinner - to spin forever ;) User might be waiting for this finished a long time

File: src/settings/BatchGroupConfigurationSettings/BatchGroupCongiurationSettingsForm.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        renderEnd={end}
+      />
+    );
+  };
+
+  return (
+    <Pane
+      data-test-batch-group-configuration-settings
+      defaultWidth="fill"
+      footer={getPaneFooter()}
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupCongiurationSettingsForm.js]: Personally, I believe calling functions at render is not the best way. Why it's a function at the first place?

File: src/settings/BatchGroupConfigurationSettings/BatchGroupCongiurationSettingsForm.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      paneTitle={paneTitle}
+    >
+      {batchGroups.length > 1
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupCongiurationSettingsForm.js]: this Select and KeyValue looks like a solid component to me. Also I tried to find form and no luck, but the PR's title says it's about `settings form`. What am I missing?

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js [hunk 8]
```diff
@@ -1,7 +1,86 @@
...
+      paneTitle={paneTitle}
+      setBatchGroupId={setBatchGroupId}
+    />
+  );
+};
+
+BatchGroupConfigurationSettings.manifest = Object.freeze({
+  batchGroups: {
+    ...batchGroupsResource,
+    accumulate: true,
```
Comment (NikitaSedyx) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: why not to add this to `batchGroupsResource`?

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js [hunk 9]
```diff
@@ -1,7 +1,86 @@
...
+        <Spinner />
+      </Pane>
+    );
+  }
+
+  return (
+    <BatchGroupConfigurationSettingsForm
+      batchGroups={batchGroups}
+      batchGroupsOptions={batchGroupsOptions}
+      initialValues={{}}
```
Comment (NikitaSedyx) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: please make a constant for initial values

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js [hunk 10]
```diff
@@ -1,7 +1,86 @@
...
+      >
+        <Spinner />
+      </Pane>
+    );
+  }
+
+  return (
+    <BatchGroupConfigurationSettingsForm
+      batchGroups={batchGroups}
+      batchGroupsOptions={batchGroupsOptions}
```
Comment (NikitaSedyx) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: you pass `batchGroups` and `batchGroupsOptions` to form, mb form should build `batchGroupsOptions`?

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js (hunk 5 above)
Comment (NikitaSedyx) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: hm, if list is empty spinner won't be shown

Comment (NikitaSedyx) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: but I agree, it's better to show it based on flag

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js (hunk 2 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: Renamed it

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js (hunk 3 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: Just for now - `onSubmit` prop is required for form.

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js (hunk 4 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: You're right, there are no strict requirements. As we discussed, I've added `LoadingPane` from `stripes-components`

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js (hunk 5 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: Added flag `isLoading`

File: src/settings/BatchGroupConfigurationSettings/BatchGroupCongiurationSettingsForm.js (hunk 6 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupCongiurationSettingsForm.js]: Ok, created component for it

File: src/settings/BatchGroupConfigurationSettings/BatchGroupCongiurationSettingsForm.js (hunk 7 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupCongiurationSettingsForm.js]: Form fields will be added in [UINV-103](https://issues.folio.org/browse/UINV-103). Select and KeyValue are out of the form.

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js (hunk 9 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: Removed it as unused

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationFormFooter.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
+import React from 'react';
+import { FormattedMessage } from 'react-intl';
+import PropTypes from 'prop-types';
+
+import {
+  Button,
+  PaneFooter,
+} from '@folio/stripes/components';
+
+const BatchGroupConfigurationFormFooter = ({
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationFormFooter.js]: does it differ so much from a common acq FormFooter?

File: src/settings/BatchGroupConfigurationSettings/BatchGroupsField.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    );
+};
+
+BatchGroupsField.propTypes = {
+  batchGroups: PropTypes.arrayOf(PropTypes.object),
+  setSelectedBatchGroupId: PropTypes.func.isRequired,
+};
+
+BatchGroupsField.defaultProps = {
+  batchGroups: [],
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupsField.js]: @folio-org/acquisitions-ui guys, I believe we need a discussion here. IMHO, we should have a rule  that  empty arrays shouldn't be default values anywhere. First reason - it hides the real value: if value was not assigned yet we can't show a spinner in a place of such component. Even if that is not a requirement for now - it's better to have this feature enabled.
Secondly, empty array doesn't equal to empty array in JS, so it's kind of performance reason. Basically, React will treat this prop as changed on every render.
Any thoughts?

File: test/bigtest/interactors/BatchGroupConfigurationSettingsInteractor.js [hunk 13]
```diff
@@ -4,4 +4,8 @@ import {
 
 export default @interactor class BatchGroupConfigurationSettingsInteractor {
   static defaultScope = '[data-test-batch-group-configuration-settings]';
+
+  whenLoaded() {
+    return this.timeout(5000).when(() => this.isLoaded);
```
Comment (aliaksei-chumakou) [inline test/bigtest/interactors/BatchGroupConfigurationSettingsInteractor.js]: but there is no `isLoaded`, as far as I can see

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationFormFooter.js (hunk 11 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationFormFooter.js]: Common `FormFooter` includes `Save` and `Cancel` buttons only

File: test/bigtest/interactors/BatchGroupConfigurationSettingsInteractor.js (hunk 13 above)
Comment (AndreiShumski2) [reply test/bigtest/interactors/BatchGroupConfigurationSettingsInteractor.js]: oops, missed it, thx

File: src/settings/BatchGroupConfigurationSettings/BatchGroupsField.js (hunk 12 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupsField.js]: Agreed. Removed default props `batchGroups`

Comment (NikitaSedyx) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupsField.js]: 1. it depends on case and requirements, but I agree it's not required here
2. I don't see any issues with array default props as defaultProps is static and it's the same array

File: test/bigtest/interactors/BatchGroupConfigurationSettingsInteractor.js [hunk 14]
```diff
@@ -1,7 +1,14 @@
 import {
   interactor,
+  isPresent,
 } from '@bigtest/interactor';
 
 export default @interactor class BatchGroupConfigurationSettingsInteractor {
   static defaultScope = '[data-test-batch-group-configuration-settings]';
+
+  isLoaded = isPresent('[data-test-pane-header-title]');
```
Comment (aliaksei-chumakou) [inline test/bigtest/interactors/BatchGroupConfigurationSettingsInteractor.js]: there are might be several panes on the page with headers. I think the idea to use unique attributes

Comment (aliaksei-chumakou) [inline test/bigtest/interactors/BatchGroupConfigurationSettingsInteractor.js]: there are might be several panes on the page with headers. I think the idea to use unique attributes

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js [hunk 15]
```diff
@@ -1,7 +1,69 @@
...
+    [selectedBatchGroupId],
+  );
+
+  if (isLoading) {
+    return (
+      <LoadingPane defaultWidth="fill" />
+    );
+  }
+
+  if (!batchGroups) {
```
Comment (aliaksei-chumakou) [inline src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: does it make sense here?

File: test/bigtest/interactors/BatchGroupConfigurationSettingsInteractor.js (hunk 14 above)
Comment (AndreiShumski2) [reply test/bigtest/interactors/BatchGroupConfigurationSettingsInteractor.js]: IMO, only one Pane with attr `data-test-batch-group-configuration-settings`, however, I added id

File: src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js (hunk 15 above)
Comment (AndreiShumski2) [reply src/settings/BatchGroupConfigurationSettings/BatchGroupConfigurationSettings.js]: removed it and set `isLoading` true

**Pull request merged**